### PR TITLE
Add New Features

### PR DESCRIPTION
**Pull request by dshanske**
List in changelog.

### CODE REVIEW

**Comment by snarfed** [comment]
woo, yay, congrats! i'll look soon. nice work!

File: micropub.php [hunk 1]
```diff
@@ -20,6 +20,23 @@
 //   curl -i -d 'code=CODE&me=SITE&client_id=indieauth&redirect_uri=https://indieauth.com/success' 'https://tokens.indieauth.com/token'
 // 4. Extract the access_token parameter from the response body.
 
+// For debugging purposes this will bypass Micropub authentication
```
Comment (snarfed) [inline micropub.php]: huh. did you notice that it doesn't require auth when it's running on localhost? given that, do you think still we want this?

File: micropub.php [hunk 2]
```diff
@@ -20,6 +20,23 @@
...
+// in favor of WordPress authentication
+// Using this to test querying(q=) parameters quickly
+if ( ! defined( 'MICROPUB_LOCAL_AUTH' ) )
+    define('MICROPUB_LOCAL_AUTH', '0');
+
+// Allows for a custom Authentication and Token Endpoint
```
Comment (snarfed) [inline micropub.php]: good idea, thanks!

File: micropub.php [hunk 3]
```diff
@@ -20,6 +20,23 @@
...
+if ( ! defined( 'MICROPUB_AUTHENTICATION_ENDPOINT' ) )
+    define('MICROPUB_AUTHENTICATION_ENDPOINT', 'https://indieauth.com/auth');
+if ( ! defined( 'MICROPUB_TOKEN_ENDPOINT' ) )
+    define('MICROPUB_TOKEN_ENDPOINT', 'https://tokens.indieauth.com/token');
+
+// For debugging purposes this will set all Micropub posts to Draft
+if ( ! defined( 'MICROPUB_DRAFT_MODE' ) )
```
Comment (snarfed) [inline micropub.php]: also a good idea!

File: micropub.php [hunk 4]
```diff
@@ -70,14 +87,30 @@ public static function parse_query($wp) {
...
+      if (!is_user_logged_in()) {
+        auth_redirect();
+      }
+      $user_id = wp_get_current_user();
+    }   
+    else {
+      $user_id = Micropub::authorize();
+    }
+    $headers = getallheaders();
+    if (isset($headers['Accept']) && $headers['Accept']=='application/json') {
```
Comment (snarfed) [inline micropub.php]: minor style nit: spaces around `==`

File: micropub.php (hunk 1 above)
Comment (dshanske) [inline micropub.php]: I actually test on a remote server. A $1 a month VPS.

File: micropub.php [hunk 5]
```diff
@@ -70,14 +87,30 @@ public static function parse_query($wp) {
...
+      $json = true;
```
Comment (snarfed) [inline micropub.php]: style nit: just do `$json = isset($headers['Accept']) && ...` instead of this whole conditional block.

Comment (snarfed) [inline micropub.php]: ...on second thought though, it looks like `$json` isn't used at all? so maybe just remove this altogether?

File: micropub.php [hunk 6]
```diff
@@ -89,11 +122,16 @@ public static function parse_query($wp) {
       $args['post_author'] = $user_id;
     }
 
-    if (!isset($_POST['url']) || $_POST['action'] == 'create') {
+    if (!isset($_POST['u']) || $_POST['action'] == 'create') {
```
Comment (snarfed) [inline micropub.php]: keep `url` also for backward compatibility?

also we need to add `u` to line 107.

File: micropub.php (hunk 5 above)
Comment (dshanske) [inline micropub.php]: I was going to add JSON support, as indicated. But didn't get to it. So I can dump the line for now.

File: micropub.php [hunk 7]
```diff
@@ -89,11 +122,16 @@ public static function parse_query($wp) {
...
       if ($user_id && !user_can($user_id, 'publish_posts')) {
         Micropub::error(403, 'user id ' . $user_id . ' cannot publish posts');
       }
-      $args['post_status'] = 'publish';
+      if (MICROPUB_DRAFT_MODE==0) {
```
Comment (snarfed) [inline micropub.php]: style nit: consider `$args['post_status'] = MICROPUB_DRAFT_MODE ? 'draft' : 'publish';`

File: micropub.php (hunk 6 above)
Comment (dshanske) [inline micropub.php]: Yes, we should have u and url for now, I believe.

File: micropub.php (hunk 7 above)
Comment (dshanske) [inline micropub.php]: Easy change.

File: micropub.php (hunk 1 above)
Comment (snarfed) [inline micropub.php]: sgtm

File: micropub.php [hunk 8]
```diff
@@ -197,6 +235,25 @@ private static function authorize() {
...
   }
 
+  private static function return_query($user_id) {
+    switch($_GET['q']) {
+      case 'syndicate-to':
+      // Fallback
+      case 'mp-syndicate-to':
+        status_header(200);
+        header('Content-type: application/x-www-form-urlencoded');
+        // Return Empty Syndication Target with Filter
```
Comment (snarfed) [inline micropub.php]: very very minor: either remove this or drop capitalization :P

File: micropub.php [hunk 9]
```diff
@@ -197,6 +235,25 @@ private static function authorize() {
     return NULL;
   }
 
+  private static function return_query($user_id) {
+    switch($_GET['q']) {
+      case 'syndicate-to':
+      // Fallback
+      case 'mp-syndicate-to':
+        status_header(200);
```
Comment (snarfed) [inline micropub.php]: minor: these two lines will probably apply to all queries, not just `syndicate-to`, so consider moving them above the switch

File: micropub.php [hunk 10]
```diff
@@ -197,6 +235,25 @@ private static function authorize() {
...
+        header('Content-type: application/x-www-form-urlencoded');
+        // Return Empty Syndication Target with Filter
+        $syndication = apply_filters('micropub_syndicate-to', array(), $user_id);
+        if (!empty($syndication)) {
+          $syndication_string = 'syndicate-to[]=' . implode('&syndicate-to[]=', $syndication);
```
Comment (snarfed) [inline micropub.php]: very minor: no need for the var, just echo the string directly

File: micropub.php (hunk 9 above)
Comment (dshanske) [inline micropub.php]: I get the first line is the header line, what is the second? The 200 return wouldn't happen with an unsupported query string.

File: micropub.php [hunk 11]
```diff
@@ -416,11 +471,9 @@ private static function check_error($result) {
    * The micropub autodicovery meta tags
    */
   public static function html_header() {
-?>
-<link rel="micropub" href="<?php echo site_url('?micropub=endpoint') ?>">
-<link rel="authorization_endpoint" href="https://indieauth.com/auth">
-<link rel="token_endpoint" href="https://tokens.indieauth.com/token">
-<?php
+    echo '<link rel="micropub" href="' . site_url('?micropub=endpoint') . '">';
+    echo '<link rel="authorization_endpoint" href="' . MICROPUB_AUTHENTICATION_ENDPOINT . '">';
```
Comment (snarfed) [inline micropub.php]: missing `$`s?

Comment (snarfed) [inline micropub.php]: oh nm, not needed since these are macros? or whatever they're called?

File: readme.txt [hunk 12]
```diff
@@ -48,9 +48,10 @@ Adds one WordPress filter, `before_micropub($wp_args)`, and one hook,
 `after_micropub($post_id)`.
 
 Delegates token handling to
-[tokens.indieauth.com](https://tokens.indieauth.com/). For ease of development,
-if the WordPress site is running on `localhost`, it logs a warning if the access
-token is missing or invalid and still allows the request.
+[tokens.indieauth.com](https://tokens.indieauth.com/) by default. For ease of
+development, if the WordPress site is running on `localhost`, it logs a warning
+if the access token is missing or invalid and still allows the request. 
+There is also a wp-config option to use WordPress authentication.
```
Comment (snarfed) [inline readme.txt]: please document the three new config options explicitly?

File: readme.txt [hunk 13]
```diff
@@ -78,6 +79,15 @@ TODO
 
 == Changelog ==
 
+= 0.4 =
+* Store all properties in post meta as opposed to predefined list
+* Support setting authentication and token endpoint in wp-config
```
Comment (snarfed) [inline readme.txt]: please mention the config options' names explicitly here too

File: micropub.php (hunk 11 above)
Comment (dshanske) [inline micropub.php]: They are constants.

File: micropub.php (hunk 6 above)
Comment (snarfed) [inline micropub.php]: still need to add `u` to line 107.

File: readme.txt [hunk 14]
```diff
@@ -44,13 +44,18 @@ Micropub properties:
 * `summary`
 * `url`
 
-Adds one WordPress filter, `before_micropub($wp_args)`, and one hook,
-`after_micropub($post_id)`.
+Adds the following filters:
+* `before_micropub($wp_args)`
+* 'micropub_syndicate-to', array(), $user_id)'
```
Comment (snarfed) [inline readme.txt]: backticks please for code in markdown, not single quotes

File: readme.txt [hunk 15]
```diff
@@ -64,6 +69,18 @@ and pull requests are welcome!
 
 Install from the WordPress plugin directory or put `micropub.php` in your plugin directory. No setup needed.
 
+== Configuration Options ==
+
+These configuration options can be enabled by adding them to your wp-config.php
+
+*  define('MICROPUB_LOCAL_AUTH', '1') - Bypasses Micropub authentication in
```
Comment (snarfed) [inline readme.txt]: very minor style nits: backticks around the code snippets please, and only one space here after `*`

**Comment by snarfed** [comment]
lgtm otherwise. thanks for your patience with me being so pedantic. :P ready to merge as soon as these last changes are in and the manual testing all works!

**Comment by snarfed** [comment]
yay, thanks for sticking with this. congrats!

**Pull request merged**